summingbird-storm/src/main/scala/com/twitter/summingbird/storm/StormPlatform.scala

```diff
@@ -204,49 +202,58 @@ abstract class Storm(options: Map[String, Options], transformConfig: Summingbird
+    hookedSpout.getSpout
+  }
+
+  private def getCountersForJob(jobId: JobId) = JobCounters.getCountersForJob(jobId).getOrElse(Nil)
+
+  /*
+  This function takes the metrics and counters to register to the TopologyContext of a JobID and outputs a function which can be called on openHook.
+   */
+  private def createRegistrationFunction(metrics: SpoutStormMetrics, counters: TraversableOnce[(Group, Name)], jobID: JobId): Externalizer[(TopologyContext) => Unit] = {
+    Externalizer({ context: TopologyContext =>
       // Register metrics passed in SpoutStormMetrics option.
       metrics.metrics().foreach {
         x: StormMetric[IMetric] =>
           context.registerMetric(x.name, x.metric, x.interval.inSeconds)
       }
       // Register summingbird counter metrics.
-      StormStatProvider.registerMetrics(jobID, context, countersForSpout)
+      StormStatProvider.registerMetrics(jobID, context, counters.toSeq)
       SummingbirdRuntimeStats.addPlatformStatProvider(StormStatProvider)
     })
-    val hookedTormentaSpout = tormentaSpout.openHook(registerAllMetrics.get)
-
-    val summerOpt: Option[SummerNode[Storm]] = stormDag.dependantsOf(node.asInstanceOf[StormNode]).collect { case s: SummerNode[Storm] => s }.headOption
-
-    val stormSpout = summerOpt match {
-      /*
-         *  As the spout is being followed by a summer, we do not have map-side aggregation handled in the spout.
-         *  This means Summer is the place where the aggregation should happen.
-         */
-      case Some(s) => createSpoutToFeedSummer(stormDag, s, hookedTormentaSpout)
-      case None => hookedTormentaSpout.getSpout
-    }
-    topologyBuilder.setSpout(nodeName, stormSpout, sourceParallelism)
   }
 
-  private def createSpoutToFeedSummer(stormDag: Dag[Storm], node: StormNode, spout: Spout[(Timestamp, Any)]) = {
+  private def createSpoutToFeedSummer[K, V](stormDag: Dag[Storm], node: StormNode, builder: SummerBuilder, spout: Spout[(Timestamp, Any)], fn: Externalizer[(TopologyContext) => Unit]): IRichSpout = {
     val summerParalellism = getOrElse(stormDag, node, DEFAULT_SUMMER_PARALLELISM)
     val summerBatchMultiplier = getOrElse(stormDag, node, DEFAULT_SUMMER_BATCH_MULTIPLIER)
     val keyValueShards = executor.KeyValueShards(summerParalellism.parHint * summerBatchMultiplier.get)
-    /*
-     This method is called only when SourceNode is followed by SummerNode.
-     The Summer exists. There will be always not more than one summer, taken care by OnlinePlan.
-    */
-    val summerProducer = node.members.collect { case s: Summer[_, _, _] => s }.head.asInstanceOf[Summer[Storm, _, _]]
+
+    val summerProducer = node.members.collect { case s: Summer[_, _, _] => s }.head.asInstanceOf[Summer[Storm, K, V]]
     val batcher = summerProducer.store.mergeableBatcher
-    val kvinj = new KeyValueInjection[Int, CMap[(_, BatchID), (Timestamp, _)]]
+    val kvinj = new KeyValueInjection[(K, BatchID), (Timestamp, V)]
```

Comment:
We are moving this logic and aggregation into the AggregateSpoutOutputCollector.